Implement map init and store for Int types

ruvi-d opened the pull request:
## Purpose
Addresses two sub tasks (Map init and store for Int member types only) listed under #89 
closes #89

ruvi-d commented:
Note that there is currently no codegen logic to deinit the Map object. Suggestions are welcome.

KavinduZoysa commented on compiler/llvmFunctionGen.cpp:
```diff
@@ -170,20 +171,20 @@ LLVMTypeRef BIRFunction::getLLVMTypeRefOfType(TypeDecl *typeD) {
     return LLVMPointerType(LLVMInt8Type(), 0);
   case TYPE_TAG_ANY:
     return LLVMPointerType(LLVMInt64Type(), 0);
+  case TYPE_TAG_MAP:
+    return LLVMPointerType(LLVMInt64Type(), 0);
```
Shall we combine these 2 cases?

ruvi-d replied:
Will do

KavinduZoysa commented on compiler/llvmFunctionGen.cpp:
```diff
@@ -169,21 +170,20 @@ LLVMTypeRef BIRFunction::getLLVMTypeRefOfType(TypeDecl *typeD) {
   case TYPE_TAG_STRING:
     return LLVMPointerType(LLVMInt8Type(), 0);
   case TYPE_TAG_ANY:
+  case TYPE_TAG_MAP:
     return LLVMPointerType(LLVMInt64Type(), 0);
```
@ruvi-d , just to know, what is the reason to set the i64* for map type? What is the theory behind that?

ruvi-d replied:
I didn't have a good reason to pick that. Ideally, I would have liked to use something akin to a `void *` so that there is some protection against accidental de-referencing.

KavinduZoysa replied:
Thanks @ruvi-d

shubhamnarlawar commented:
@ruvi-d - This PR looks fine to me!

shubhamnarlawar commented:
@ayonam - Can you please take quick look at this PR to give any suggestions if required?

ramakotareddy commented:
@ruvi-d  i have gone through Translation side code, Code looks good to me.
@ayonam Can you please give your comments.

ruvi-d commented:
@shubhamnarlawar and @ramakotareddy noted thanks. Please use the Github option "Review changes" and then select "Approve / Request changes" to complete your review

shubhamnarlawar reviewed (approved):
Good to merge!

KavinduZoysa commented:
`map<int> marks = {}`
@ruvi-d , I got the .ll file for above simple code and it is as followed.

```
define void @main() {
entry:
  %"%0" = alloca i32, align 4
  %"%1" = alloca i64*, align 8
  %"%2" = alloca i32, align 4
  br label %bb0

bb0:                                              ; preds = %entry
  %0 = call i64* @map_new_int()
  store i64* %0, i64** %"%1", align 8
  br label %bb2

bb1:                                              ; No predecessors!
  br label %bb2

bb2:                                              ; preds = %bb1, %bb0
  ret void
}
```

Here we assign the returned value to `%0`. I think it will be more clear if we can assign it to temp value like %1_temp and store it in %1, WDYT?

ruvi-d commented:
> Here we assign the returned value to `%0`. I think it will be more clear if we can assign it to temp value like %1_temp and store it in %1, WDYT?

I am a bit torn about that. You are right that it would improve readability, but it is a redundant step that would likely get optimized out when we add LLVM IR optimization passes later on. 

@ayonam : any thoughts?

KavinduZoysa commented:
> > Here we assign the returned value to `%0`. I think it will be more clear if we can assign it to temp value like %1_temp and store it in %1, WDYT?
> 
> I am a bit torn about that. You are right that it would improve readability, but it is a redundant step that would likely get optimized out when we add LLVM IR optimization passes later on.
> 
> @ayonam : any thoughts?

Yes, redundancy point of view, I agree with you.

ruvi-d commented:
Summary of open question to @ayonam 

1. There is currently no codegen logic to deinit (free) the Map object. Is there a recommended approach to call deinit for the local variables at least?
2. To hold the map pointer returned from Rust side, is there a preference between using *i8 or *i64 on the codegen side? 
3. When we are generating LLVM IR code, is it better to use more temp variables and be verbose? or is it better to use as few as possible and assign directly to local variables?

ayonam commented on compiler/llvmMap.cpp:
```diff
@@ -0,0 +1,65 @@
+  // Find VarDecl corresponding to lhs to determine member type
+  VarDecl *lhsVar = funcObj->getNameVarDecl(lhsName);
+  assert(lhsVar);
+  assert(lhsVar->getVarKind() == LOCAL_VAR_KIND);
```
Why is this assertion?  Can't the LHS var be a global variable?

ruvi-d replied:
will check for global variables and get back.

ayonam commented on compiler/llvmMap.cpp:
```diff
@@ -0,0 +1,65 @@
+  if (!mapStoreFunc)
+    mapStoreFunc = getMapIntStoreDeclaration(modRef, pkgObj);
+
+  LLVMValueRef lhsOpTempRef = funcObj->getLocalToTempVar(getLhsOperand());
```
I think you should check the scope of LHS operand and then either call this getLocalToTempVar() if it is a local variable or else do something like a GEP (Global Element Pointer).

ruvi-d replied:
noted. will check with global variables

ayonam commented on compiler/llvmMap.cpp:
```diff
@@ -0,0 +1,65 @@
+  LLVMValueRef lhsOpTempRef = funcObj->getLocalToTempVar(getLhsOperand());
+
+  LLVMValueRef rhsOpRef = funcObj->getLocalVarRefUsingId(rhsName);
+  LLVMValueRef keyRef = funcObj->getLocalToTempVar(keyOp);
```
Ditto for the RHS op and the KEY op.

ruvi-d replied:
noted. thanks

ayonam commented on compiler/llvmMap.cpp:
```diff
@@ -0,0 +1,65 @@
+  LLVMTypeRef funcType = LLVMFunctionType(LLVMVoidType(), paramTypes, 3, 0);
+  LLVMValueRef addedFuncRef =
+      LLVMAddFunction(modRef, "map_store_int", funcType);
+  pkg->addArrayFunctionRef("map_store_int", addedFuncRef);
```
This is fine for now.  But going forward, as we add different kinds of Maps with different data types, should we be having polymorphic functions on the Rust side or do we have different names for each data type?  If it is the former, then what should this function return?  Probably it should return nothing.  It should simply setup the types of each of the polymorphic functions and return void.

ayonam commented on compiler/llvmStructure.cpp:
```diff
@@ -0,0 +1,65 @@
+  string lhsName = getLhsOperand()->name();
+  LLVMValueRef lhsOpRef = funcObj->getLocalVarRefUsingId(lhsName);
+  if (!lhsOpRef)
+    lhsOpRef = pkgObj->getGlobalVarRefUsingId(lhsName);
```
Here you are checking for both local as well as global var refs.  You need to do like this everywhere.

ruvi-d replied:
noted. thanks

ayonam commented on compiler/llvmStructure.cpp:
```diff
@@ -0,0 +1,65 @@
+  // Only handle Int type
+  if (memberTypeTag != TYPE_TAG_INT) {
+    std::cerr << "Non INT type maps are currently not supported" << std::endl;
+    assert(false);
```
You should call llvm_unreachable() instead of assert(false)

ayonam commented on compiler/llvmStructure.cpp:
```diff
@@ -0,0 +1,65 @@
+  LLVMValueRef addedFuncRef = LLVMAddFunction(modRef, "map_new_int", funcType);
+  pkg->addArrayFunctionRef("map_new_int", addedFuncRef);
+  return addedFuncRef;
+}
```
Again the same comment here regarding polymorphic functions on the Rust side to handle initialization of maps of different data types.

ayonam commented on test/map/mapBasic.bal:
```diff
@@ -0,0 +1,7 @@
+    map<int> marks = {};
+    marks["jon"] = 60;
+}
+// CHECK: length=1
```
Shouldn't there also be a check to see if the marks["jon"] is indeed 60 or not?

ruvi-d replied:
That will be handled as par of #128

ruvi-d commented on compiler/BIR.h:
```diff
@@ -65,7 +65,7 @@ enum InstructionKind {
   INSTRUCTION_KIND_MOVE = 20,
   INSTRUCTION_KIND_CONST_LOAD = 21,
   INSTRUCTION_KIND_NEW_STRUCTURE,
-  INSTRUCTION_KIND_NEW_MAP,
```
There is no "new map" instruction as per the BIR spec listed here: https://github.com/ballerina-platform/ballerina-lang/blob/bd8e12ac24a70079247d1b178de888cb0e76a96a/docs/bir-spec/src/main/resources/kaitai/bir.ksy#L988

We get `new struct` instruction with a LHS value of type Map for new Maps

ruvi-d commented on compiler/BIRReader.cpp:
```diff
@@ -139,23 +140,35 @@ bool ConstantPoolSet::getBooleanCp(uint32_t index) {
+    assert(shapeEntry->getTag() ==
+           ConstantPoolEntry::tagEnum::TAG_ENUM_CP_ENTRY_SHAPE);
+    ShapeCpInfo *typeShapeCp = static_cast<ShapeCpInfo *>(shapeEntry);
+    int memeberType = typeShapeCp->getTypeTag();
```
Thanks for spotting it. Will fix.

ruvi-d commented on compiler/llvmFunctionGen.cpp:
```diff
@@ -169,21 +170,20 @@ LLVMTypeRef BIRFunction::getLLVMTypeRefOfType(TypeDecl *typeD) {
     return LLVMPointerType(LLVMInt8Type(), 0);
   case TYPE_TAG_ANY:
+  case TYPE_TAG_MAP:
     return LLVMPointerType(LLVMInt64Type(), 0);
```
Please see this question regarding this topic: https://github.com/ballerina-platform/nballerina/pull/102#issuecomment-785677115
We were not sure whether it is best to use a int-8 or int-64 pointer to hold the address returned from the Rust runtime.

ruvi-d replied:
I've pushed a change to use int8* instead of int64*

ruvi-d commented on compiler/BIR.h:
```diff
@@ -754,8 +795,7 @@ class BIRFunction : public BIRNode {
   VarDecl *getReceiver() { return receiver; }
   Param *getRestParam() { return restParam; }
-  vector<VarDecl *> getLocalVars() { return localVars; }
-  VarDecl *getLocalVar(int i) { return localVars[i]; }
```
We are not using them anymore

ruvi-d commented on compiler/BIR.h:
```diff
@@ -544,12 +582,15 @@ class TypeDescInsn : public NonTerminatorInsn {
 private:
   Operand *rhsOp;
+  void mapInsnTranslate(VarDecl *lhsVar, LLVMModuleRef &modRef);
+  LLVMValueRef getNewMapIntDeclaration(LLVMModuleRef &modRef, BIRPackage *pkg);
```
The BIR for new Map is actually a `new structure` with a LHS of type Map. So I am having to handle the new map Code gen from the StrutureIsn class. Hope that makes sense?

ruvi-d commented on compiler/BIR.h:
```diff
@@ -777,9 +817,10 @@ class BIRFunction : public BIRNode {
-  void setLocalVar(VarDecl *var) { localVars.push_back(var); }
+  void setLocalVar(string name, VarDecl *var) {
+    localVars.insert(pair<string, VarDecl *>(name, var));
+  }
```
Yes makes sense. Fill fix.

ruvi-d commented on compiler/llvmMap.cpp:
```diff
@@ -0,0 +1,65 @@
+#include "BIR.h"
+
+// new Map Instruction and Codegen logic are in the llvmStructure.cpp
```
There is no `new map` BIR instruction. What we are getting is a `new structure` instruction withe the RHS that has a Map type. That is why I am handling the codegen for new map as part of the new structure instruction class

ruvi-d commented on compiler/llvmMap.cpp:
```diff
@@ -0,0 +1,65 @@
+MapStoreInsn::MapStoreInsn(Location *pos, InstructionKind kind, Operand *lOp,
+                           Operand *KOp, Operand *rOp)
+    : NonTerminatorInsn(pos, kind, lOp), keyOp(KOp), rhsOp(rOp) {}
+
```
You are only supposed to add inline method definitions to the header right?

ruvi-d commented:
Addressed the review comments from Ayonam. The discussion about why the new map instruction is handled instead the new Structure instruction is unresolved.

ayonam commented:
@ruvi-d   @KavinduZoysa 

Please find my replies inlined below:

> Summary of open question to @ayonam
> 
>     1. There is currently no codegen logic to deinit (free) the Map object. Is there a recommended approach to call deinit for the local variables at least?
> 

Local variables that are created as part of LLVM generated code will automatically get deallocated once the current function returns and we need not do anything about it.  For those that are created by the Rust side of things through an explicit allocation on the heap, we need to deallocate them explicitly (or through a memory manager / garbage collector).  So although these are local variables from the programmer's perspective, they are actually allocated by Rust on the heap.  

>     2. To hold the map pointer returned from Rust side, is there a preference between using *i8 or *i64 on the codegen side?

Actually since we treat that pointer as an opaque pointer, it is equivalent to void* in the C/C++ paradigm.  Now in LLVM there is no concept of a void pointer.  A void pointer is represented as an i8*.  So I think we should treat such map pointers as i8*.

> 
>     3. When we are generating LLVM IR code, is it better to use more temp variables and be verbose? or is it better to use as few as possible and assign directly to local variables?

Don't bother about the temp variables.  The optimizer downstream is going to straighten them out.  Currently we are not turning on the optimizer (by passing -O0) and hence you still see all those temps in the LLVM dump.  We are doing so, so that we can debug things more easily.  Once we reach certain level of maturity in our codegen, we will turn on the optimizer and then you should see most of these temps going away.

ruvi-d commented:
Noted.. Thanks for the answers Ayonam

ayonam commented:
> Addressed the review comments from Ayonam. The discussion about why the new map instruction is handled instead the new Structure instruction is unresolved.

@Ruvi, Do we have a resolution to this one as yet?  If yes, then this should be good to approve.  If not, then I'd prefer if you can open a separate Issue about this and we can discuss it there.

ruvi-d commented:
> > Addressed the review comments from Ayonam. The discussion about why the new map instruction is handled instead the new Structure instruction is unresolved.
> 
> @ruvi, Do we have a resolution to this one as yet? If yes, then this should be good to approve. If not, then I'd prefer if you can open a separate Issue about this and we can discuss it there.

No resolution yet. I've started a slack thread on it too here: https://ballerina-platform.slack.com/archives/C01LNSAHK18/p1614654420035700
Let's close this PR and I will create an Github issue to track it separately.
